src/pyprojroot/criterion.py

```diff
@@ -9,35 +9,50 @@
 import typing
 from os import PathLike as _PathLike
 
+
+_PathType = typing.Union[_PathLike, str]
+_CriterionType = typing.Union[
+    typing.Callable[[_PathType], bool],
+    typing.Callable[[_pathlib.Path], bool],
+    _PathType,
+    _pathlib.Path,
+    typing.Iterable[typing.Callable[[_PathType], bool]],
+    typing.Iterable[typing.Callable[[_pathlib.Path], bool]],
+]
+
 # TODO: It would be nice to have a class that encapsulates these checks,
 # so that we can implement methods like |, !, &, ^ operators
 
 # TODO: Refactor in a way that allows creation of reasons
 
 
-def as_root_criterion(criterion) -> typing.Callable:
+def as_root_criterion(
+    criterion: _CriterionType,
+) -> typing.Callable[[_pathlib.Path], bool]:
     if callable(criterion):
         return criterion
 
     # criterion must be a Collection, rather than just Iterable
     if isinstance(criterion, _PathLike):
-        criterion = [criterion]
-    criterion = list(criterion)
+        criterion_collection = [criterion]
+    else:
+
+        criterion_collection = list(criterion)  # type: ignore[arg-type]
```

Comment:
For anyone that's interested, this needs to be converted into a list here from a generic iterable since we need to be able to restart the iterator.